In [ ]:
#импортируем библиотеки
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import pandas as pd

In [ ]:
# загружаем набор данных
data = pd.read_csv('/content/diabetes.csv')
data


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [ ]:
X = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

X = torch.tensor(X, dtype=torch.float32)
y = torch.tensor(y, dtype=torch.float32).reshape(-1, 1)

X

tensor([[  6.0000, 148.0000,  72.0000,  ...,  33.6000,   0.6270,  50.0000],
        [  1.0000,  85.0000,  66.0000,  ...,  26.6000,   0.3510,  31.0000],
        [  8.0000, 183.0000,  64.0000,  ...,  23.3000,   0.6720,  32.0000],
        ...,
        [  5.0000, 121.0000,  72.0000,  ...,  26.2000,   0.2450,  30.0000],
        [  1.0000, 126.0000,  60.0000,  ...,  30.1000,   0.3490,  47.0000],
        [  1.0000,  93.0000,  70.0000,  ...,  30.4000,   0.3150,  23.0000]])

3 слоя

In [ ]:
# определяем модель с 3 внутренними слоями
model_3_layers = nn.Sequential(
    nn.Linear(8, 16),
    nn.ReLU(),
    nn.Linear(16, 16),
    nn.ReLU(),
    nn.Linear(16, 16),
    nn.ReLU(),
    nn.Linear(16, 8),
    nn.ReLU(),
    nn.Linear(8, 1),
    nn.Sigmoid()
)

In [ ]:
# тренируем модель с 3 внутренними слоями
loss_fn = nn.BCELoss()
optimizer = optim.Adam(model_3_layers.parameters(), lr=0.001)

In [ ]:
n_epochs = 100
batch_size = 10

for epoch in range(n_epochs):
    for i in range(0, len(X), batch_size):
        Xbatch = X[i:i+batch_size]
        y_pred = model_3_layers(Xbatch)
        ybatch = y[i:i+batch_size]
        loss = loss_fn(y_pred, ybatch)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    print(f'Итерация № {epoch}, latest loss {loss}')

Итерация № 0, latest loss 0.6401985883712769
Итерация № 1, latest loss 0.6414381265640259
Итерация № 2, latest loss 0.6316633820533752
Итерация № 3, latest loss 0.6253698468208313
Итерация № 4, latest loss 0.6195770502090454
Итерация № 5, latest loss 0.6050199270248413
Итерация № 6, latest loss 0.581944465637207
Итерация № 7, latest loss 0.5610040426254272
Итерация № 8, latest loss 0.5550487637519836
Итерация № 9, latest loss 0.5589388608932495
Итерация № 10, latest loss 0.5566352009773254
Итерация № 11, latest loss 0.5538742542266846
Итерация № 12, latest loss 0.5454301834106445
Итерация № 13, latest loss 0.5565385222434998
Итерация № 14, latest loss 0.5468841791152954
Итерация № 15, latest loss 0.5397754907608032
Итерация № 16, latest loss 0.540846586227417
Итерация № 17, latest loss 0.5301909446716309
Итерация № 18, latest loss 0.5259948968887329
Итерация № 19, latest loss 0.5272554755210876
Итерация № 20, latest loss 0.5318195223808289
Итерация № 21, latest loss 0.5154401063919067


In [ ]:
# высчитываем accuracy для модели с 3 внутренними слоями
with torch.no_grad():
    y_pred = model_3_layers(X)
accuracy_3_layers = (y_pred.round() == y).float().mean()
print(f"Accuracy модели с 3 внутренними слоями: {accuracy_3_layers}")

Accuracy модели с 3 внутренними слоями: 0.76171875


4 слоя

In [ ]:
# модель с 4 внутренними слоями
model_4_layers = nn.Sequential(
    nn.Linear(8, 16),
    nn.ReLU(),
    nn.Linear(16, 16),
    nn.ReLU(),
    nn.Linear(16, 16),
    nn.ReLU(),
    nn.Linear(16, 16),
    nn.ReLU(),
    nn.Linear(16, 8),
    nn.ReLU(),
    nn.Linear(8, 1),
    nn.Sigmoid()
)

In [ ]:
# тренируем модель с 4 внутренними слоями
loss_fn = nn.BCELoss()
optimizer = optim.Adam(model_4_layers.parameters(), lr=0.001)

In [ ]:
for epoch in range(n_epochs):
    for i in range(0, len(X), batch_size):
        Xbatch = X[i:i+batch_size]
        y_pred = model_4_layers(Xbatch)
        ybatch = y[i:i+batch_size]
        loss = loss_fn(y_pred, ybatch)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    print(f'Итерация № {epoch}, latest loss {loss}')

Итерация № 0, latest loss 0.561241865158081
Итерация № 1, latest loss 0.5230149030685425
Итерация № 2, latest loss 0.5013667941093445
Итерация № 3, latest loss 0.48417767882347107
Итерация № 4, latest loss 0.46543359756469727
Итерация № 5, latest loss 0.4607119560241699
Итерация № 6, latest loss 0.44434189796447754
Итерация № 7, latest loss 0.4368549585342407
Итерация № 8, latest loss 0.43047571182250977
Итерация № 9, latest loss 0.433137446641922
Итерация № 10, latest loss 0.43292319774627686
Итерация № 11, latest loss 0.4253876209259033
Итерация № 12, latest loss 0.4281640350818634
Итерация № 13, latest loss 0.4244318902492523
Итерация № 14, latest loss 0.42852291464805603
Итерация № 15, latest loss 0.4321548640727997
Итерация № 16, latest loss 0.43108701705932617
Итерация № 17, latest loss 0.4358142614364624
Итерация № 18, latest loss 0.4342385232448578
Итерация № 19, latest loss 0.4421977400779724
Итерация № 20, latest loss 0.44166067242622375
Итерация № 21, latest loss 0.440344691

In [ ]:
# высчитываем accuracy для модели с 4 внутренними слоями
with torch.no_grad():
    y_pred = model_4_layers(X)
accuracy_4_layers = (y_pred.round() == y).float().mean()
print(f"Accuracy модели с 4 внутренними слоями: {accuracy_4_layers}")


Accuracy модели с 4 внутренними слоями: 0.8177083134651184


Результат: точность у модели с 4 внутренними слоями (0.8177083134651184) выше,чем у модели с тремя (0.76171875)